File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/15990-1, 2, 3 더하기 5/solution.py
import heapq
import sys
si = sys.stdin.readline
MOD = 1000000009

dy = [[0, 0, 0, 0] for _ in range(100005)]
# 초기값 채우기
dy[1][1] = 1
dy[2][2] = 1
dy[3][1] = 1
dy[3][2] = 1
dy[3][3] = 1

# 점화식을 토대로 dy 배열 채우기
for i in range(4, 100001):
    for cur in range(1, 4):
        for prev in range(1, 4):
            if cur == prev: continue
            dy[i][cur] += dy[i - cur][prev]
            dy[i][cur] %= MOD

T = int(si())
for _ in range(T):
    print(sum(dy[int(si())]) % MOD)

File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/15990-1, 2, 3 더하기 5/solution.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#pragma warning(disable:4996)
using namespace std;
#define NM 100005
#define MOD 1000000009

int Dy[NM][4];

void preprocess() {
	// 초기값 구하기
	Dy[1][1] = 1;
	Dy[2][2] = 1;
	Dy[3][1] = Dy[3][2] = Dy[3][3] = 1;

	// 점화식을 토대로 Dy 배열 채우기
	for (int i = 4; i <= 100000; i++) {
		for (int cur = 1; cur <= 3; cur++) {
			for (int prev = 1; prev <= 3; prev++) {
				if (cur == prev) continue;
				Dy[i][cur] += Dy[i - cur][prev];
				Dy[i][cur] %= MOD;
			}
		}
	}
}

void pro() {
	int T; cin >> T;
	while (T--) {
		int n; cin >> n;
		int ans = 0;
		for (int cur = 1; cur <= 3; cur++) {
			ans += Dy[n][cur];
			ans %= MOD;
		}
		cout << ans << "\n";
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	preprocess();
	pro();
	return 0;
}